pkg/scheduler/scheduler.go

```diff
@@ -319,6 +319,40 @@ func (s *Scheduler) nominate(ctx context.Context, workloads []workload.Info, sna
 	return entries
 }
 
+// resourcesToReserve calculates how much of the available resources in cq/cohort assignment should be reserved.
+func resourcesToReserve(e *entry, cq *cache.ClusterQueue) cache.FlavorResourceQuantities {
+	if e.assignment.RepresentativeMode() != flavorassigner.Preempt {
+		return e.assignment.Usage
+	}
+	reservedUsage := make(cache.FlavorResourceQuantities)
+	for flavor, resourceUsage := range e.assignment.Usage {
+		for resource, usage := range resourceUsage {
+			if _, ok := reservedUsage[flavor]; !ok {
+				reservedUsage[flavor] = make(map[corev1.ResourceName]int64)
+			}
+			rg := cq.RGByResource[resource]
+			cqQuota := cache.ResourceQuota{}
+			for _, cqFlavor := range rg.Flavors {
+				if cqFlavor.Name == flavor {
+					cqQuota = *cqFlavor.Resources[resource]
+					break
+				}
+			}
+			if cq.Usage[flavor][resource] <= cqQuota.Nominal {
+				// since cq and assignment doesn't borrow, cq usage can't exceed nominal quota.
+				reservedUsage[flavor][resource] = min(usage, cqQuota.Nominal-cq.Usage[flavor][resource])
+			} else if cqQuota.BorrowingLimit != nil {
+				// if cq is borrowing, cq usage can't exceed nominal + borrowing limit
+				reservedUsage[flavor][resource] = min(usage, cqQuota.Nominal+*cqQuota.BorrowingLimit-cq.Usage[flavor][resource])
```

Comment:
We don't want to block up to the borrowing limit. We only want to block up to nominal.

Reply:
@mimowo what do you think should we do if borrowing with preemption? I guess we should block all of the usage in that case?

Reply:
Updated, we shouldn't go over cqQuota.Nominal+*cqQuota.BorrowingLimit if borrowing limit is set

Reply:
> @mimowo what do you think should we do if borrowing with preemption? I guess we should block all of the usage in that case?

TBH, I'm not sure, I'm wondering if we can simplify 
```
if cqQuota.BorrowingLimit == nil {
					reservedUsage[flavor][resource] = usage
				} else {
					reservedUsage[flavor][resource] = min(usage, cqQuota.Nominal+*cqQuota.BorrowingLimit-cq.Usage[flavor][resource])
				}
```
to just 
```
reservedUsage[flavor][resource] = usage
```
@yaroslava-serdiuk do you know a scenario when this is needed? 

This looks odd to me at the moment, because in the scenario of single shared CQ as [here])https://github.com/kubernetes-sigs/kueue/tree/main/keps/1337-preempt-within-cohort-while-borrowing#story-1) when `cqQuota.Nominal=0`, then if `borrowingLimit=null` we reserve `usage`, but if `borrowingLimit=nominal for the shared (essentially the same a null)`, and `cq.Usage[flavor][resource]=borrowigLimit`, then we reserve `0`.

Reply:
I tested that the existing integration tests pass if we do `reservedUsage[flavor][resource] = usage` when `a.Borrowing=true`. Still, I'm not 100% sure what the logic should be in case assignment needs borrowing, but this PR changed logic for `a.Borrowing=true` without an integration test for this scenario.

Reply:
Had discussion with @mimowo offline. I explained the condition:
```if cqQuota.BorrowingLimit == nil {
  reservedUsage[flavor][resource] = usage
} else {
  reservedUsage[flavor][resource] = min(usage, cqQuota.Nominal+*cqQuota.BorrowingLimit-cq.Usage[flavor][resource])
}```
However this is unlikely will be used, since the borrowing limit is often nil.
I also constructed example when the condition unblock another workload from borrowing: 
![Untitled Jam 1](https://github.com/kubernetes-sigs/kueue/assets/43413443/3ee2a4f6-d8ee-4bb4-96d8-2e9662e8869c)

Reply:
Thank you for the sync and explanations indeed. 

One thing that it let me understand is that the StrictFIFO isn't really the main point here (as the title suggests). If the queue has only one workload it does not matter what is its type. Maybe we could say in the release notes:
"Fix a bug that a preempting workload in one CQ could block another pending workload in another CQ even if the other workload could be scheduled by borrowing".

Regarding the extra condition I think it adds complexity (and the code is already complex). The scenario isn't tested currently by integration tests. Also, the importance of the scenario from the user perspective seems questionable, because it is only relevant if `borrowingLimit<nominalQuota of shared CQ`. I could open a PR to simplify. WDYT @alculquicondor ?

Reply:
>  Also, the importance of the scenario from the user perspective seems questionable, because it is only relevant if borrowingLimit<nominalQuota of shared CQ. 

@mimowo, Could you clarify this situation? Which does this mean, `a: borrowingLimit<(nominalQuota of shared CQ)` or `b: (borrowingLimit<nominalQuota) of shared CQ`?

If you mean `a`, I think `a` will often happen. I guess admins want to set `borrowingLimit<(nominalQuota of shared CQ)` in each ClusterQueue since admins want to prevent situations where one of the CQ monopolizes all shared capacity.

If my understanding is incorrect , please let me know.

Reply:
> @mimowo, Could you clarify this situation? Which does this mean, `a: borrowingLimit<(nominalQuota of shared CQ)` or `b: (borrowingLimit<nominalQuota) of shared CQ`?

I meant (a) `borrowingLimit<(nominalQuota of shared CQ)`. 

> If you mean a, I think a will often happen. I guess admins want to set borrowingLimit<(nominalQuota of shared CQ) in each ClusterQueue since admins want to prevent situations where one of the CQ monopolizes all shared capacity.

Makes sense, I was thinking about the scenario as in [here](https://github.com/kubernetes-sigs/kueue/tree/main/keps/1337-preempt-within-cohort-while-borrowing#story-1), but both are relevant. 

So the remaining question: do we care about the scenario, that a workload that is borrowing (B) could get blocked by another workload (A) that is preempting and borrowing in the cohort. The trade-off does not seem obviousm because letting the other workload run may cause the waiting workload wait longer.

Reply:
Feel free to open a PR